pandas/core/strings.py

```diff
@@ -1736,11 +1712,259 @@ def cons_row(x):
+        'a b d'
+
+        By default, NA values in the Series are ignored. Using `na_rep`, they
+        can be given a representation:
+
+        >>> s.str.cat(sep=' ', na_rep='?')
+        'a b ? d'
+
+        If `others` is specified, corresponding values are concatenated with
+        the separator. Result will be a Series of strings.
+
+        >>> s.str.cat(['A', 'B', 'C', 'D'], sep=',')
+        0    a,A
+        1    b,B
+        2    NaN
+        3    d,D
+        dtype: object
+
+        Missing values will remain missing in the result, but can again be
+        represented using `na_rep`
+
+        >>> s.str.cat(['A', 'B', 'C', 'D'], sep=',', na_rep='-')
+        0    a,A
+        1    b,B
+        2    -,C
+        3    d,D
+        dtype: object
+
+        If `sep` is not specified, the values are concatenated without
+        separation.
+
+        >>> s.str.cat(['A', 'B', 'C', 'D'], na_rep='-')
+        0    aA
+        1    bB
+        2    -C
+        3    dD
+        dtype: object
+
+        Series with different indexes can be aligned before concatenation. The
+        `join`-keyword works as in other methods.
+
+        >>> t = pd.Series(['d', 'a', 'e', 'c'], index=[3, 0, 4, 2])
+        >>> s.str.cat(t, join=None, na_rep='-')
+        0    ad
+        1    ba
+        2    -e
+        3    dc
+        dtype: object
+        >>>
+        >>> s.str.cat(t, join='left', na_rep='-')
+        0    aa
+        1    b-
+        2    -c
+        3    dd
+        dtype: object
+        >>>
+        >>> s.str.cat(t, join='outer', na_rep='-')
+        0    aa
+        1    b-
+        2    -c
+        3    dd
+        4    -e
+        dtype: object
+        >>>
+        >>> s.str.cat(t, join='inner', na_rep='-')
+        0    aa
+        2    -c
+        3    dd
+        dtype: object
+        >>>
+        >>> s.str.cat(t, join='right', na_rep='-')
+        3    dd
+        0    aa
+        4    -e
+        2    -c
+        dtype: object
+
+        For more examples, see :ref:`here <text.concatenate>`.
+        """
+        from pandas.core.series import Series
+        from pandas.core.frame import DataFrame
+        from pandas.core.reshape.concat import concat
+        from collections import Iterable
+        from itertools import chain
+
+        if isinstance(others, str):
+            raise ValueError("Did you mean to supply a `sep` keyword?")
+
+        # If others is not one of the types we can deal with, but still
+        # iterable, turn into list. Needed to check its contents below.
+        if (not isinstance(others, (list, Series, DataFrame, np.ndarray))
+                and isinstance(others, Iterable)):
+            others = list(others)
+
+        if join is None and (isinstance(others, (Series, DataFrame))
+                             or (isinstance(others, list)
+                                 and any(isinstance(x, (Series, DataFrame))
+                                         for x in others))):
```

Comment:
you are adding a ton of logic here and its pretty nested. break this out into some sub-functions. it will be much easier to read and follow. clearly separate the purpose of ``str_cat`` from what you are doing here. The embeded list case great complicated all of this code. This should either be really simple of simply dropped.

Reply:
I didn't advocate the nested list case, @jorisvandenbossche did - I'm guessing mainly because it's currently legal to use list of list-likes. I just added the capability to align on top of that possibility, and with all the possible cases, it's not simple. I would prefer disallowing the list-of case, but not sure that's the best way to go.

Reply:
And the logic you commented at around the warning is simply to avoid emitting it in cases where non of the objects in others has an index.

Reply:
Thanks for the suggestion of breaking it into a subfunction - the code got much easier now.
There's effectively no nesting anymore. Could you have another look?